propagation/propagation.go

```diff
@@ -69,14 +70,31 @@ func (c MapCarrier) Keys() []string {
 // HeaderCarrier adapts http.Header to satisfy the TextMapCarrier interface.
 type HeaderCarrier http.Header
 
-// Get returns the value associated with the passed key.
+// Get returns the first value associated with the passed key. The key is
+// case-sensitive and will only return the value that matches exactly. If there
+// are no values associated with the key, Get returns "".
 func (hc HeaderCarrier) Get(key string) string {
-	return http.Header(hc).Get(key)
+	// Use the underlying map to avoid normalising the key and thereby
+	// changing its value.
+	if hc == nil {
+		return ""
+	}
+	value, ok := http.Header(hc)[key]
+	if !ok || len(value) == 0 {
+		return ""
+	}
+
+	return value[0]
 }
 
-// Set stores the key-value pair.
+// Set sets the header entries associated with key to the single element value.
+// It replaces any existing values associated with key.
+//
+// The key is case-sensitive, it is used exactly as passed.
 func (hc HeaderCarrier) Set(key string, value string) {
-	http.Header(hc).Set(key, value)
+	// Use the underlying map to set the value to avoid normalising the key and
+	// thereby changing its value.
+	http.Header(hc)[strings.ToLower(key)] = []string{value}
```

Comment:
Does this also need to check whether `hc == nil`?

I'm confused about the comment that says the value is set in the map directly to avoid normalization, but the key is passed through `strings.ToLower()` before use.

Reply:
the code and the comment might have drifted as I tried different options, however the conclusion I got to was that it was not possible with the current API to achieve what I wanted because of the reasons mentioned here: https://github.com/open-telemetry/opentelemetry-go/pull/3049#issuecomment-1204474240 I think this can be closed as it's not a valid solution at the moment